File: src/main/java/com/owncloud/android/utils/ClipboardUtil.java
/*
 * Nextcloud Android client application
 *
 * @author Andy Scherzinger
 * Copyright (C) 2018 Andy Scherzinger
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU AFFERO GENERAL PUBLIC LICENSE
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU AFFERO GENERAL PUBLIC LICENSE for more details.
 *
 * You should have received a copy of the GNU Affero General Public
 * License along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package com.owncloud.android.utils;

import android.app.Activity;
import android.content.ClipData;
import android.content.ClipboardManager;
import android.content.Context;
import android.widget.Toast;

import com.owncloud.android.R;
import com.owncloud.android.lib.common.utils.Log_OC;

/**
 * Helper implementation to copy a string into the system clipboard.
 */
public final class ClipboardUtil {
    private static final String TAG = ClipboardUtil.class.getName();

    private ClipboardUtil() {
    }

    public static void copyToClipboard(Activity activity, String text) {
        copyToClipboard(activity, text, true);
    }

    public static void copyToClipboard(Activity activity, String text, boolean showToast) {
        if (text != null && text.length() > 0) {
            try {
                ClipData clip = ClipData.newPlainText(
                        activity.getString(
                                R.string.clipboard_label, activity.getString(R.string.app_name)),
                        text
                );
                ((ClipboardManager) activity.getSystemService(Context.CLIPBOARD_SERVICE)).setPrimaryClip(clip);

                if (showToast) {
                    Toast.makeText(activity, R.string.clipboard_text_copied, Toast.LENGTH_SHORT).show();
                }
            } catch (Exception e) {
                Toast.makeText(activity, R.string.clipboard_unexpected_error, Toast.LENGTH_SHORT).show();
                Log_OC.e(TAG, "Exception caught while copying to clipboard", e);
            }
        } else {
            Toast.makeText(activity, R.string.clipboard_no_text_to_copy, Toast.LENGTH_SHORT).show();
        }
    }
}

File: app/src/main/java/com/owncloud/android/utils/ClipboardUtil.kt
/*
 * Nextcloud - Android Client
 *
 * SPDX-FileCopyrightText: 2024 Alper Ozturk <alper.ozturk@nextcloud.com>
 * SPDX-FileCopyrightText: 2018 Andy Scherzinger <info@andy-scherzinger.de>
 * SPDX-License-Identifier: AGPL-3.0-or-later OR GPL-2.0-only
 */
package com.owncloud.android.utils

import android.app.Activity
import android.content.ClipData
import android.content.ClipDescription
import android.content.ClipboardManager
import android.content.Context
import android.os.Build
import android.os.PersistableBundle
import android.text.TextUtils
import android.widget.Toast
import androidx.work.Data
import androidx.work.OneTimeWorkRequestBuilder
import androidx.work.WorkManager
import com.nextcloud.client.jobs.clipboard.ClipboardClearWorker
import com.nextcloud.utils.mdm.MDMConfig
import com.owncloud.android.R
import com.owncloud.android.lib.common.utils.Log_OC
import java.util.concurrent.TimeUnit

object ClipboardUtil {
    private val TAG = ClipboardUtil::class.java.name
    private const val WORKER_CLEAR_DELAY = 15L

    @JvmStatic
    @JvmOverloads
    @Suppress("TooGenericExceptionCaught")
    fun copyToClipboard(activity: Activity, text: String?, showToast: Boolean = true) {
        if (!MDMConfig.clipBoardSupport(activity)) {
            return
        }

        if (TextUtils.isEmpty(text)) {
            Toast.makeText(activity, R.string.clipboard_no_text_to_copy, Toast.LENGTH_SHORT).show()
            return
        }

        try {
            val clipboardLabel = activity.getString(R.string.clipboard_label, activity.getString(R.string.app_name))
            val clip = getClipData(clipboardLabel, text)
            val clipboardManager = (activity.getSystemService(Context.CLIPBOARD_SERVICE) as ClipboardManager)
            clipboardManager.setPrimaryClip(clip)

            if (showToast) {
                Toast.makeText(activity, R.string.clipboard_text_copied, Toast.LENGTH_SHORT).show()
            }

            scheduleClipboardClearWorker(activity, text)
        } catch (e: Exception) {
            Toast.makeText(activity, R.string.clipboard_unexpected_error, Toast.LENGTH_SHORT).show()
            Log_OC.e(TAG, "Exception caught while copying to clipboard", e)
        }
    }

    private fun getClipData(clipboardLabel: String, text: String?): ClipData {
        return ClipData.newPlainText(clipboardLabel, text).apply {
            description.extras = PersistableBundle().apply {
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
                    putBoolean(ClipDescription.EXTRA_IS_SENSITIVE, true)
                } else {
                    putBoolean("android.content.extra.IS_SENSITIVE", true)
                }
            }
        }
    }

    private fun scheduleClipboardClearWorker(context: Context, text: String?) {
        val data = Data.Builder()
            .putString(ClipboardClearWorker.CLIPBOARD_TEXT, text)
            .build()

        val clearWorkRequest = OneTimeWorkRequestBuilder<ClipboardClearWorker>()
            .setInitialDelay(WORKER_CLEAR_DELAY, TimeUnit.SECONDS)
            .setInputData(data)
            .build()

        WorkManager.getInstance(context).enqueue(clearWorkRequest)
    }
}
